### DISCUSSION TITLE
Problem with SparseTensor in EdgeConv

### DISCUSSION THREAD

**Discussion by ikvara** [danielegrattarola/spektral#328] [Q&A]
Hi, I'm trying to fit a simple model with EdgeConv layer on dataset in Mixed mode and I get an error with the SparseTensor.

Here is my implementation:
`class StressMixedModeDataset(spektral.data.Dataset):
    def __init__(self, data_type, **kwargs):
        self.a = None
        self.data_type = data_type
        super().__init__(**kwargs)

    def read(self):
        if self.data_type=='train':
            with open('train_mixed.pkl', "rb") as fr:   
                output = pckl.load(fr)
        elif self.data_type=='val':
            with open('val_mixed.pkl', "rb") as fr:   
                output = pckl.load(fr)

        self.a = output[0]
        node_feats = output[1]
        edge_feats = output[2]
        y_labels = output[3]

        GraphList = []
        for x, y in zip(node_feats, y_labels):
          x = np.asarray(x)
          y = np.asarray(y)

          GraphList.append(spektral.data.graph.Graph(x=x, e=edge_feats, y=y))

        return GraphList

class Net(tf.keras.models.Model):
    def __init__(self):
        super().__init__()

        self.conv = spektral.layers.EdgeConv(args.n_channels)
        self.output_layer = tf.keras.layers.Dense(6)

    def call(self, inputs):
        x, a, e = inputs

        print(type(a))
        x = self.conv([x, a])
        outs = self.output_layer(x)

        return outs

def main(args):
    train = StressMixedModeDataset('train')
    val = StressMixedModeDataset('val')

    print(type(train.a))
    print(type(val.a))

    loader_tr = spektral.data.loaders.MixedLoader(train, batch_size=args.batch_size)
    loader_va = spektral.data.loaders.MixedLoader(val, batch_size=args.batch_size)

    model = Net()
    model.compile(optimizer=tf.optimizers.Adam(), loss = 'mse')

    epochs = args.epochs 
    model.fit(loader_tr.load(), 
            validation_data = loader_va.load(),
            steps_per_epoch=loader_tr.steps_per_epoch, 
            validation_steps=loader_va.steps_per_epoch,
            epochs=epochs,
            verbose=2,
            )
        
if __name__ == "__main__":
    args = parser.parse_args([] if "__file__" not in globals() else None)
    main(args)`

The adjacency matrix is a sparse matrix and is thus converted to SparseTensor. However, from what it prints I see that somewhere it gets broken. 
` <class 'scipy.sparse.coo.coo_matrix'>
<class 'scipy.sparse.coo.coo_matrix'>
<class 'tensorflow.python.framework.sparse_tensor.SparseTensor'>
Epoch 1/10
<class 'tensorflow.python.framework.ops.Tensor'>`

Then I get the following error: 
`File "/usr/local/lib/python3.7/dist-packages/spektral/layers/convolutional/message_passing.py", line 105, in call  *
      x, a, e = self.get_inputs(inputs)
  File "/usr/local/lib/python3.7/dist-packages/spektral/layers/convolutional/message_passing.py", line 181, in get_inputs  *
      assert K.is_sparse(a), "A must be a SparseTensor"     

  AssertionError: A must be a SparseTensor`

Can you please help? What am I doing wrong?

**Comment by ikvara** [comment]
Oh, I see where the problem was. The MixedLoader requires different approach to training. I'm sorry for bothering!

**Reply by danielegrattarola** [reply]
Hey no problem, I'm glad you were able to solve the issue

cheers
